Simplenote/src/main/java/com/automattic/simplenote/PreferencesFragment.java

```diff
@@ -456,6 +481,101 @@ public int compare(String text1, String text2) {
+                    importPlaintextFile(exportContents);
+                    break;
+                default:
+                    Toast.makeText(requireContext(), getString(R.string.import_unknown), Toast.LENGTH_SHORT).show();
+                    return;
+            }
+            Toast.makeText(requireContext(), getString(R.string.import_message_success), Toast.LENGTH_SHORT).show();
+        } catch (Exception e) {
+            Toast.makeText(requireContext(), getString(R.string.import_message_failure), Toast.LENGTH_SHORT).show();
+        }
+    }
+
+    private void importPlaintextFile(String exportContents) {
+        Simplenote currentApp = (Simplenote) requireActivity().getApplication();
+        Bucket<Note> noteBucket = currentApp.getNotesBucket();
+        Note note = noteBucket.newObject();
+        note.setContent(exportContents);
+        note.setCreationDate(Calendar.getInstance());
+        note.setModificationDate(note.getCreationDate());
+        note.save();
+    }
+
+    private void importJsonExport(String exportContents) throws Exception {
+        JSONObject jsonData = new JSONObject(exportContents);
+        JSONArray activeNotes = jsonData.getJSONArray("activeNotes");
+        JSONArray trashedNotes = jsonData.getJSONArray("trashedNotes");
+        List<String> trashedIds = new ArrayList<>();
+        JSONArray notesArray = new JSONArray();
+        for (int i = 0; i < activeNotes.length(); i++) {
+            notesArray.put(activeNotes.getJSONObject(i));
+        }
+        for (int j = 0; j < trashedNotes.length(); j++) {
+            notesArray.put(trashedNotes.getJSONObject(j));
+            trashedIds.add(trashedNotes.getJSONObject(j).getString("id"));
+        }
+        for (int k = 0; k < notesArray.length(); k++) {
+            JSONObject noteJson = notesArray.getJSONObject(k);
+            noteFromJson(noteJson, trashedIds).save();
+        }
+    }
+
+    private Note noteFromJson(JSONObject noteJson, List<String> trashedIds) throws Exception {
+        Simplenote currentApp = (Simplenote) requireActivity().getApplication();
+        Bucket<Note> noteBucket = currentApp.getNotesBucket();
+        Note note = noteBucket.newObject();
+        note.setContent(noteJson.has("content") ? noteJson.getString("content") : "");
+        note.setCreationDate(noteJson.has("creationDate") ? DateTimeUtils.getDateCalendar(noteJson.getString("creationDate")) : Calendar.getInstance());
+        note.setModificationDate(noteJson.has("lastModified") ? DateTimeUtils.getDateCalendar(noteJson.getString("lastModified")) : Calendar.getInstance());
+        note.setTags(noteJson.has("tags") ? jsonToList(noteJson.getJSONArray("tags")) : new ArrayList<String>());
+        note.setPinned(noteJson.has("pinned") && noteJson.getBoolean("pinned"));
+        note.setMarkdownEnabled(noteJson.has("markdown") && noteJson.getBoolean("markdown"));
+        note.setDeleted(noteJson.has("id") && trashedIds.contains(noteJson.getString("id")));
```

Comment:
this is going to have some issues as I don't know that we can or should rely on imported notes containing an id. I believe that it's not really necessary anyway, more a byproduct of the way we've chosen to build an intermediate list of notes on import.

instead we could traverse each category of notes once and avoid needing to pass in `List<String> trashedIds`. there are a couple ways we can convey the trashed status but one easy way to do it would having to think about it much is to manually alter the JSON or the output note.

```java
for (int i = 0; i < activeNotes.length(); i++) {
	noteFromJson(notesArray.getJSONObject(i)).save();
}

for (int i = 0; i < trashedNotes.length(); i++) {
	Note note = noteFromJson(notesArray.getJSONObject(i));
	note.setDeleted(true);
	note.save();
}
```

Reply:
Yes, I had doubts about this piece of code, so much better, thanks!